zerver/models/users.py

```diff
@@ -930,6 +934,8 @@ def get_user_by_delivery_email(email: str, realm: "Realm") -> UserProfile:
         "realm",
         "realm__can_access_all_users_group",
         "realm__can_access_all_users_group__named_user_group",
+        "realm__can_create_public_channel_group",
+        "realm__can_create_public_channel_group__named_user_group",
```

Comment:
These calls to `select_related` are mainly needed when returning the setting value in `/register` response. So, would probably needed for all the settings we add in future.

Reply:
Strategically, I think if we only need them for `/register`, we should just have `/register` re-select the user with extra foreign keys, rather than having every other API endpoint that works with a user fetch these.

Probably worth a `#backend` conversation, but one theory is that we could have a way to declare an API endpoint with say `prefetch_user_groups=["can_access_all_users_group"]`, and have that plumbed through to how the user object is fetched.

But probably for the vast majority of endpoints that aren't performance-sensitive, it's OK for these to get extra queries, so I think a better place to start would be to just have the `/register` code path refetch its user object with all the extra columns that it'll need?

Reply:
We need settings like can_access_all_users_group, the setting to restrict DM, etc. in the user_profile objects in endpoint for sending messages, etc. But for settings like this one, except for sending the values in /register response, they would only be used while creating streams and I guess we could afford an extra query there.

So, yeah what we can do here is that just re-fetch the realm object and call select_related in the query to fetch all the settings.

Reply:
Sounds like a plan for this detail.